tensorboard/plugins/pr_curve/summary.py

```diff
@@ -0,0 +1,190 @@
+    # We have to do this after weights multiplication since weights broadcast.
+    shape = (-1, num_classes)
+    predictions = tf.reshape(predictions, shape)
+    true_labels = tf.reshape(true_labels, shape)
+    false_labels = tf.reshape(false_labels, shape)
+
+    # To compute TP/FP/TN/FN, we are measuring a classifier
+    #   C(t) = (predictions >= t)
+    # at each threshold 't'. So we have
+    #   TP(t) = sum( C(t) * true_labels )
+    #   FP(t) = sum( C(t) * false_labels )
+    #
+    # But, computing C(t) requires computation for each t. To make it fast,
+    # observe that C(t) is a cumulative integral, and so if we have
+    #   thresholds = [t_0, ..., t_{n-1}];  t_0 < ... < t_{n-1}
+    # where n = num_thresholds, and if we can compute the bucket function
+    #   B(i) = Sum( (predictions == t), t_i <= t < t{i+1} )
+    # then we get
+    #   C(t_i) = sum( B(j), j >= i )
+    # which is the reversed cumulative sum in tf.cumsum().
+    #
+    # We can compute B(i) efficiently by taking advantage of the fact that
+    # our thresholds are evenly distributed, in that
+    #   width = 1.0 / (num_thresholds - 1)
+    #   thresholds = [0.0, 1*width, 2*width, 3*width, ..., 1.0]
+    # Given a prediction value p, we can map it to its bucket by
+    #   bucket_index(p) = floor( p * (num_thresholds - 1) )
+    # so we can use tf.scatter_add() to update the buckets in one pass.
+
+    # First compute the bucket indices for each prediction value.
+    bucket_indices = (
+        tf.cast(tf.floor(predictions * (num_thresholds-1)), tf.int32))
+    # Adjust indices by classes. For performance and simplicity, we keep
+    # the buckets (see below) as 1D array representing the tp/fp buckets for
+    # a (num_classes, num_thresholds) tensors.
+    # So, for each index in bucket_indices, its real index into this 1D array is
+    #   index + (num_thresholds * class_index)
+    class_indices = tf.reshape(tf.range(num_classes), (1, num_classes))
+    bucket_indices += num_thresholds * class_indices
+
+    with tf.name_scope('variables'):
+      # Now create the variables which correspond to the bucket values.
+      # These are flat arrays with num_thresholds valuewchargins per class.
+      tp_buckets_v = tf.Variable(
+          tf.zeros([num_classes * num_thresholds], dtype=dtype),
+          name='tp_buckets',
+          trainable=False,
+          collections=[tf.GraphKeys.LOCAL_VARIABLES])
+      fp_buckets_v = tf.Variable(
+          tf.zeros([num_classes * num_thresholds], dtype=dtype),
+          name='fp_buckets',
+          trainable=False,
+          collections=[tf.GraphKeys.LOCAL_VARIABLES])
+
+    initialize_bucket_counts = tf.variables_initializer(
+        [tp_buckets_v, fp_buckets_v])
+    with tf.control_dependencies([initialize_bucket_counts]):
+      # Create the non-flat views with class_index as first dimension.
+      tp_buckets = tf.reshape(tp_buckets_v, (num_classes, -1))
```

Comment:
The reshapes offer a different view into the data in which the first dimension is class. The previous shape is a flat continuous array.